File: java_client/src/main/java/com/cisc468share/net/ConsentManager.java
package com.cisc468share.net;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Decouples consent prompts from background TCP threads.
 *
 * The TCP router runs in a background thread and cannot safely call
 * Scanner.nextLine() while the CLI is also blocked on Scanner.nextLine()
 * in the main thread — the first keypress goes to whichever caller
 * happens to win the race.
 *
 * Fix: the background thread queues a request here and blocks on
 * the response queue. The main CLI loop polls hasPending(), prints
 * the prompt itself, reads the response, and calls respond().
 */
public class ConsentManager {

    public static class ConsentRequest {
        public final String peerName;
        public final String filename;
        public final long   filesize;

        ConsentRequest(String peerName, String filename, long filesize) {
            this.peerName = peerName;
            this.filename = filename;
            this.filesize = filesize;
        }
    }

    private final BlockingQueue<ConsentRequest> pending  = new ArrayBlockingQueue<>(1);
    private final BlockingQueue<Boolean>        response = new ArrayBlockingQueue<>(1);
    private final AtomicBoolean                 waiting  = new AtomicBoolean(false);

    /**
     * Called from a background thread.
     * Submits the request and blocks until the main thread responds.
     */
    public boolean request(String peerName, String filename, long filesize)
            throws InterruptedException {
        pending.put(new ConsentRequest(peerName, filename, filesize));
        waiting.set(true);
        return response.take();   // blocks until CLI loop calls respond()
    }

    /** True when a consent request is queued and waiting for a response. */
    public boolean hasPending() {
        return waiting.get();
    }

    /** Called from the main thread to retrieve the queued request. */
    public ConsentRequest popPending() throws InterruptedException {
        waiting.set(false);
        return pending.take();
    }

    /** Called from the main thread after the user answers. */
    public void respond(boolean accepted) throws InterruptedException {
        response.put(accepted);
    }
}

File: python_client/net/consent_manager.py
"""
ConsentManager — decouples consent prompts from background threads.

The TCP router runs in a background thread. Python's input() cannot be
safely called from two threads at once — the CLI's input() and the router's
input() fight over stdin, causing the first keypress to be consumed by the
wrong caller.

Fix: the background thread queues the consent request and blocks on a
response queue. The main CLI loop detects the pending request, prints the
prompt itself (from the main thread), reads the response, and unblocks the
background thread.
"""
import threading
import queue


class ConsentManager:

    def __init__(self):
        self._pending  = queue.Queue()   # (peer_name, filename, filesize)
        self._response = queue.Queue()   # bool: True=accepted, False=denied
        self._event    = threading.Event()

    def request(self, peer_name: str, filename: str, filesize: int) -> bool:
        """
        Called from a background thread.
        Submits the consent request and blocks until the main thread responds.
        """
        self._pending.put((peer_name, filename, filesize))
        self._event.set()
        return self._response.get()   # blocks until CLI loop calls respond()

    def is_pending(self) -> bool:
        return self._event.is_set()

    def pop_pending(self):
        """Called from the main thread to retrieve the queued request."""
        item = self._pending.get_nowait()
        self._event.clear()
        return item   # (peer_name, filename, filesize)

    def respond(self, accepted: bool):
        """Called from the main thread after the user answers."""
        self._response.put(accepted)
